userapi/src/main/java/fk/prof/userapi/util/StackLineParser.java

```diff
@@ -0,0 +1,122 @@
+package fk.prof.userapi.util;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** A utility class which parses and converts Java names from their corresponding JVM Types signatures
+ - from string of format className#methodName (argTypes)retType
+ * Created by rohit.patiyal on 26/04/17.
+ */
+public class StackLineParser {
+    private static Map<Character, String> signToPrimitiveMap = new HashMap<Character, String>(){
+        {
+            put('V',"void");
+            put('Z',"boolean");
+            put('B',"byte");
+            put('C',"char");
+            put('S',"short");
+            put('I',"int");
+            put('J',"long");
+            put('F',"float");
+            put('D',"double");
+        }
+    };
+
+    private static List<JVMTypeSignatureParser> parsers = new ArrayList<JVMTypeSignatureParser>(){
+        {
+            add(StackLineParser::parseTerminalType);
+            add(StackLineParser::parseMethodName);
+            add(StackLineParser::parseArgs);
+            add(StackLineParser::parseRet);
+        }
+    } ;
+
+    public static String convertJVMTypeSignToJava(String text){
+        StringBuilder sb = new StringBuilder();
+        int textIndex = 0;
+        int sbLength = 0;
+        text = text.replace('/','.');
+        for(JVMTypeSignatureParser parser : parsers){
```

Comment:
Why a list of parsers and trying all of them one by one. Seems like a code smell. There ideally should be a single parser which appropriately determines which parsing method to call depending on how the string starts (think of it as determining initial state in a state machine). It can also throw up error if type signature did not start with an expected value (maybe should always start with L since that is how method name starts?)